x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java

```diff
@@ -242,14 +245,54 @@ static ClusterState removeRoutingToUnassignableNodes(ClusterState currentState)
                         routedNodeIdsToRemove
                     )
                 );
-                TrainedModelAssignment.Builder assignmentBuilder = TrainedModelAssignment.Builder.fromAssignment(assignment);
-                routedNodeIdsToRemove.forEach(assignmentBuilder::removeRoutingEntry);
+
+                TrainedModelAssignment.Builder assignmentBuilder = removeRoutingBuilder(
+                    routedNodeIdsToRemove,
+                    currentState.metadata().nodeShutdowns().getAllNodeIds(),
+                    assignment
+                );
+
                 builder.updateAssignment(assignment.getDeploymentId(), assignmentBuilder.calculateAndSetAssignmentState());
             }
         }
         return update(currentState, builder);
     }
 
+    private static TrainedModelAssignment.Builder removeRoutingBuilder(
+        Set<String> nodeIds,
+        Set<String> shuttingDownNodes,
+        TrainedModelAssignment assignment
+    ) {
+        TrainedModelAssignment.Builder assignmentBuilder = TrainedModelAssignment.Builder.fromAssignment(assignment);
+
+        for (String nodeIdToRemove : nodeIds) {
+            if (shuttingDownNodes.contains(nodeIdToRemove) == false) {
+                logger.debug(
+                    () -> format("[%s] Removing route for unassignable node id [%s]", assignment.getDeploymentId(), nodeIdToRemove)
+                );
+
+                assignmentBuilder.removeRoutingEntry(nodeIdToRemove);
+            } else if (assignment.getNodeRoutingTable()
+                .get(nodeIdToRemove)
```

Comment:
I believe it'll be guaranteed to exist because the ids come from the routing table for the assignment passed in already from this line:

https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java#L236

```
Set<String> routedNodeIdsToRemove = Sets.difference(assignment.getNodeRoutingTable().keySet(), assignableNodes);
```

Can't hurt to add a null check though so I'll add one 👍